### PR TITLE
Making serialization functions specific instead of generic and removing helper code.

### PR DESCRIPTION
**Pull request by xvzcf**
- I've removed the generic `serialize` and `deserialize` functions that take in the number of bits to use as a parameter, and replaced them by specific functions that work on 1, 4, 10, and 12 bits only
- This meant the BitVector code was no longer needed, so I've removed that too

### CODE REVIEW

File: src/kem/kyber768/ind_cpa.rs [hunk 1]
```diff
@@ -284,28 +284,26 @@ pub(crate) fn decrypt(
...
     );
 
     // sˆ := Decode_12(sk)
-    let mut secret_as_ntt_ring_element_bytes = secret_key.chunks(BITS_PER_RING_ELEMENT / 8);
+    let mut secret_as_ntt_ring_element_bytes = secret_key.chunks(BYTES_PER_RING_ELEMENT);
     for i in 0..secret_as_ntt.len() {
-        secret_as_ntt[i] = deserialize_little_endian(
-            12,
+        secret_as_ntt[i] = deserialize_little_endian_12(
             secret_as_ntt_ring_element_bytes.next().expect("secret_as_ntt_ring_element_bytes should have enough bytes to deserialize to secret_as_ntt"),
```
Comment (franziskuskiefer) [inline src/kem/kyber768/ind_cpa.rs:301]: We should be able to use `secret_key.chunks(BYTES_PER_RING_ELEMENT).enumerate()` for the loop now and get rid of the unwrap here.

File: src/kem/kyber768/ind_cpa.rs [hunk 2]
```diff
@@ -133,55 +179,13 @@ pub(crate) fn generate_keypair(
...
-        // 2 sampling coins * 64
-        let prf_output: [u8; 128] = PRF(&prf_input);
-
-        let r = sample_from_binomial_distribution_with_2_coins(prf_output);
-        r_as_ntt[i] = ntt_representation(r);
-    }
-    (r_as_ntt, domain_separator)
-}
-
 fn encode_and_compress_u(input: [KyberPolynomialRingElement; RANK]) -> Vec<u8> {
```
Comment (franziskuskiefer) [inline src/kem/kyber768/ind_cpa.rs:182]: We should be able to return an array here, right?
Then we can get rid of another vector and don't use the `collect` + `as_array` in `encrypt`.

File: src/kem/kyber768/ind_cpa.rs [hunk 3]
```diff
@@ -194,11 +198,10 @@ pub(crate) fn encrypt(
...
 ) -> Result<CiphertextCpa, BadRejectionSamplingRandomnessError> {
     // tˆ := Decode_12(pk)
-    let mut t_as_ntt_ring_element_bytes = public_key.chunks(BITS_PER_RING_ELEMENT / 8);
+    let mut t_as_ntt_ring_element_bytes = public_key.chunks(BYTES_PER_RING_ELEMENT);
     let mut t_as_ntt = [KyberPolynomialRingElement::ZERO; RANK];
     for i in 0..t_as_ntt.len() {
-        t_as_ntt[i] = deserialize_little_endian(
-            12,
+        t_as_ntt[i] = deserialize_little_endian_12(
             t_as_ntt_ring_element_bytes.next().expect(
```
Comment (franziskuskiefer) [inline src/kem/kyber768/ind_cpa.rs:205]: Couldn't we use the iterator on the `public_key` here as well?

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
Thanks!
I left a few comments for some more improvements. If you want to push some of them into followups, just mark them as such.

Please run rustfmt on all the code.

File: src/kem/kyber768/serialize.rs [hunk 4]
```diff
@@ -1,76 +1,140 @@
...
+    for (i, chunk) in re.coefficients.chunks_exact(2).enumerate() {
+        let coefficient1 = chunk[0].value as u8;
+        let coefficient2 = chunk[1].value as u8;
+
+        serialized[i] = (coefficient2 << 4) | coefficient1;
     }
 
     serialized
 }
+pub fn deserialize_little_endian_4(
```
[the remark this answers is not in the document]
Comment (xvzcf) [reply src/kem/kyber768/serialize.rs:48]: I was thinking we'd add some documentation like: "this file contains implementations of `serialize_n` and `deserialize_n` for various values of n relevant to the Kyber spec", and then describe what they do in terms of the generic parameter `n`.

File: src/kem/kyber768/sampling.rs [hunk 5]
```diff
@@ -37,9 +37,7 @@ pub fn sample_from_uniform_distribution(
     Err(BadRejectionSamplingRandomnessError)
 }
 
-pub fn sample_from_binomial_distribution_with_2_coins(
-    randomness: [u8; 128],
-) -> KyberPolynomialRingElement {
+pub fn sample_from_binomial_distribution_2(randomness: [u8; 128]) -> KyberPolynomialRingElement {
```
[the remark this answers is not in the document]
Comment (xvzcf) [reply src/kem/kyber768/sampling.rs:40]: We don't, I just thought I'd leave it in there to provide an extra bit of documentation.

File: src/kem/kyber768/ind_cpa.rs (hunk 3 above)
Comment (xvzcf) [reply src/kem/kyber768/ind_cpa.rs]: Yup, changed it in `ind_cpa.rs` whereever I could.

File: src/kem/kyber768/serialize.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (franziskuskiefer) [reply src/kem/kyber768/serialize.rs:48]: Sounds good. Can you do that here?

File: src/kem/kyber768/sampling.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (franziskuskiefer) [reply src/kem/kyber768/sampling.rs:40]: You can add a comment describing what you do, then the `_2` is fine 🙂 We need a comment anyway.

**Review by franziskuskiefer — APPROVED** [review]
Thanks.
I'm approving but please add some comments. When the comments are resolved you can merge it.

**Pull request merged**